sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -485,6 +509,105 @@ case class GpuCast(
     }
   }
 
+  private def castStructToString(input: GpuColumnVector,
+    legacyCastToString: Boolean, inputSchema: Array[StructField]): ColumnVector = {
+    // The brackets that are used in casting structs and maps to strings
+    val (leftBracket, rightBracket) = if (legacyCastToString) ("[", "]") else ("{", "}")
+    var separatorColumn: ColumnVector = null
+    var spaceColumn: ColumnVector = null
+    val columns: ArrayBuffer[ColumnVector] = new ArrayBuffer[ColumnVector]()
+    // coreColumns tracks the casted child columns
+    val coreColumns: ArrayBuffer[ColumnVector] = new ArrayBuffer[ColumnVector]()
+
+    try {
+      withResource(GpuScalar.from(leftBracket, StringType)) { bracketScalar =>
+        columns += ColumnVector.fromScalar(bracketScalar, input.getRowCount().toInt)
+      }
+      withResource(input.getBase().getChildColumnView(0)) { childView =>
+        withResource(childView.copyToColumnVector()) { childVector =>
```

Comment:
Why do we need to copy this to a column vector? I understand because of the APIs we have right now it is needed, but it would be nice to avoid the copy if possible.

Reply:
Ideally we would not copy this column vector, which is why I had been using `toString` in earlier iterations. How might I avoid this copy?

Reply:
The signature of `doColumnar` takes a `GpuColumnVector` It uses the `GpuColumnVector` to extract its Spark data type so it can do the cast. A `ColumnVector` is a `ColumnView`. As such I would propose that we insert another API in between that we could use for recursive calls.

```
override def doColumnar(input: GpuColumnVector): ColumnVector =
    doColumnar(input.getBase, input.dataType())

private def doColumnar(input: ColumnView, sparkType: DataType): COlumnVector = {
   (sparkType, dataType) match {
      ...
   }
}
```

Then you can call the new doColumnar without making  copy of the data.

```
withResource(input.getBase().getChildColumnView(0)) { childView =>
  columns += doColumnar(childView, inputSchema(0).dataType)
}
```

Reply:
There a few things around decimal casting that didn't translate nicely. In some cases they want to do a zero-copy, so I've filtered them out in the `GpuColumnVector` wrapper API. In those cases, if those cases are found when doing a `ColumnView` cast, it's copied to a ColumnVector. Reworking the decimal support to be `ColumnView` friendly version can be done in a follow up PR if necessary, but it seemed a bit out of scope and was non-trivial.

Reply:
Could we have a follow on issue to clean this up? The issue appears to be because of incRefCount, and there should be ways for us to work around that without too much difficulty.

Reply:
#2043